src/Engine/Component/AnimationController.cpp

```diff
@@ -75,9 +124,77 @@ void Component::AnimationController::UpdateAnimation(float deltaTime) {
         glm::quat finalRot = glm::lerp(rotation1, rotation2, interpolation);
 
         glm::mat4 matrixRot = glm::mat4(finalRot);
-        matrixRot[3][3] = 1.0f;
 
         skeleton->skeletonBones[i]->globalTx = skeleton->skeletonBones[skeleton->skeletonBones[i]->parentId]->globalTx * skeleton->skeletonBones[i]->localTx * matrixRot;
         bones[i] = skeleton->skeletonBones[i]->globalTx * skeleton->skeletonBones[i]->inversed;
     }
 }
+
+void AnimationController::Interpolate(float deltaTime) {
+    auto size = activeAction1->animationClip->animation->numBones > activeAction2->animationClip->animation->numBones ? activeAction2->animationClip->animation->numBones : activeAction1->animationClip->animation->numBones;
+
+    if (activeAction1->animationClip->animation->numBones != bonesToInterpolate1.size()) {
+        bonesToInterpolate1.clear();
+        bonesToInterpolate1.shrink_to_fit();
+        bonesToInterpolate1.resize(activeAction1->animationClip->animation->numBones);
+    }
+
+    if (activeAction2->animationClip->animation->numBones != bonesToInterpolate2.size()) {
+        bonesToInterpolate2.clear();
+        bonesToInterpolate2.shrink_to_fit();
+        bonesToInterpolate2.resize(activeAction2->animationClip->animation->numBones);
+    }
+
+    for (auto i = 0; i < size; ++i) {
+        float interpolation = activeTransition->transitionProcess / activeTransition->transitionTime;
+
+        // Clamp interpolation.
+        if (interpolation > 0.999f)
+            interpolation = 0.999f;
+        else if (interpolation < 0.001f)
+            interpolation = 0.001f;
+
+        // Convert to quaternion to interpolate animation then back to matrix.
+        glm::quat rotation1 = glm::quat_cast(bonesToInterpolate1[i]);
+        glm::quat rotation2 = glm::quat_cast(bonesToInterpolate2[i]);
+        glm::quat finalRot = glm::lerp(rotation1, rotation2, interpolation);
+
+        glm::mat4 matrixRot = glm::mat4(finalRot);
+        
+        bones[i] = matrixRot;
+    }
+}
+
+void AnimationController::SetBool(std::string name, bool value) {
+    for (auto i = 0; i < controller->boolMap.size(); ++i) {
+        if (strcmp(name.c_str(), controller->boolMap[i]->name) == 0) {
+            controller->boolMap[i]->value = value;
+            return;
+        }
+    }
+}
+
+void AnimationController::SetFloat(std::string name, float value) {
+    for (auto i = 0; i < controller->floatMap.size(); ++i) {
+        if (strcmp(name.c_str(), controller->floatMap[i]->name) == 0) {
+            controller->floatMap[i]->value = value;
+            return;
+        }
+    }
+}
+
+bool AnimationController::GetBool(std::string name) {
+    for (auto i = 0; i < controller->boolMap.size(); ++i) {
+        if (strcmp(name.c_str(), controller->boolMap[i]->name) == 0) {
+            return controller->boolMap[i]->value;
+        }
+    }
+}
+
+float AnimationController::GetFloat(std::string name) {
```

Comment:
Could be const?

Reply:
`const` itself doesn't do anything (the value is copied), but const ref is useful.